File: csharp/src/Ice/Internal/TraceUtil.cs
// Copyright (c) ZeroC, Inc.

using System.Diagnostics;
using System.Globalization;

namespace Ice.Internal;

internal sealed class TraceUtil
{
    internal static void traceSend(
        OutputStream str,
        Instance instance,
        ConnectionI connection,
        Logger logger,
        TraceLevels tl)
    {
        if (tl.protocol >= 1)
        {
            int p = str.pos();
            var iss = new Ice.InputStream(instance, str.getEncoding(), str.getBuffer(), false);
            iss.pos(0);

            using (var s = new System.IO.StringWriter(CultureInfo.CurrentCulture))
            {
                byte type = printMessage(s, iss, connection);

                logger.trace(tl.protocolCat, "sending " + getMessageTypeAsString(type) + " " + s.ToString());
            }
            str.pos(p);
        }
    }

    internal static void traceRecv(Ice.InputStream str, ConnectionI connection, Ice.Logger logger, TraceLevels tl)
    {
        if (tl.protocol >= 1)
        {
            int p = str.pos();
            str.pos(0);

            using (var s = new System.IO.StringWriter(CultureInfo.CurrentCulture))
            {
                byte type = printMessage(s, str, connection);

                logger.trace(tl.protocolCat, "received " + getMessageTypeAsString(type) + " " + s.ToString());
            }
            str.pos(p);
        }
    }

    internal static void trace(string heading, InputStream str, ConnectionI connection, Logger logger, TraceLevels tl)
    {
        if (tl.protocol >= 1)
        {
            int p = str.pos();
            str.pos(0);

            using (var s = new StringWriter(CultureInfo.CurrentCulture))
            {
                s.Write(heading);
                printMessage(s, str, connection);

                logger.trace(tl.protocolCat, s.ToString());
            }
            str.pos(p);
        }
    }

    private static readonly HashSet<string> slicingIds = new();

    internal static void traceSlicing(string kind, string typeId, string slicingCat, Ice.Logger logger)
    {
        lock (_globalMutex)
        {
            if (slicingIds.Add(typeId))
            {
                using var s = new StringWriter(CultureInfo.CurrentCulture);
                s.Write("unknown " + kind + " type `" + typeId + "'");
                logger.trace(slicingCat, s.ToString());
            }
        }
    }

    private static void printIdentityFacetOperation(System.IO.StringWriter s, Ice.InputStream str)
    {
        try
        {
            Ice.ToStringMode toStringMode = str.instance().toStringMode();

            var identity = new Ice.Identity(str);
            s.Write("\nidentity = " + Ice.Util.identityToString(identity, toStringMode));

            string[] facet = str.readStringSeq();
            s.Write("\nfacet = ");
            if (facet.Length > 0)
            {
                s.Write(Ice.UtilInternal.StringUtil.escapeString(facet[0], "", toStringMode));
            }

            string operation = str.readString();
            s.Write("\noperation = " + operation);
        }
        catch (System.IO.IOException)
        {
            Debug.Assert(false);
        }
    }

    private static void printRequest(System.IO.StringWriter s, Ice.InputStream str)
    {
        int requestId = str.readInt();
        s.Write("\nrequest id = " + requestId);
        if (requestId == 0)
        {
            s.Write(" (oneway)");
        }

        printRequestHeader(s, str);
    }

    private static void printBatchRequest(System.IO.StringWriter s, Ice.InputStream str)
    {
        int batchRequestNum = str.readInt();
        s.Write("\nnumber of requests = " + batchRequestNum);

        for (int i = 0; i < batchRequestNum; ++i)
        {
            s.Write("\nrequest #" + i + ':');
            printRequestHeader(s, str);
        }
    }

    private static void printReply(System.IO.StringWriter s, Ice.InputStream str)
    {
        int requestId = str.readInt();
        s.Write("\nrequest id = " + requestId);

        // We can't (shouldn't) use the generated code to unmarshal a possibly unknown reply status.
        var replyStatus = (ReplyStatus)str.readByte();
        s.Write("\nreply status = ");
        s.Write(replyStatus);

        switch (replyStatus)
        {
            case ReplyStatus.Ok:
            case ReplyStatus.UserException:
                EncodingVersion v = str.skipEncapsulation();
                if (v != Ice.Util.Encoding_1_0)
                {
                    s.Write("\nencoding = ");
                    s.Write(Ice.Util.encodingVersionToString(v));
                }
                break;

            case ReplyStatus.ObjectNotExist:
            case ReplyStatus.FacetNotExist:
            case ReplyStatus.OperationNotExist:
                printIdentityFacetOperation(s, str);
                break;

            default:
                string message = str.readString();
                s.Write("\nmessage = ");
                s.Write(message);
                break;
        }
    }

    private static void printRequestHeader(System.IO.StringWriter s, Ice.InputStream str)
    {
        printIdentityFacetOperation(s, str);

        try
        {
            byte mode = str.readByte();
            s.Write("\nmode = " + ((OperationMode)mode).ToString());

            int sz = str.readSize();
            s.Write("\ncontext = ");
            while (sz-- > 0)
            {
                string key = str.readString();
                string val = str.readString();
                s.Write(key + '/' + val);
                if (sz > 0)
                {
                    s.Write(", ");
                }
            }

            Ice.EncodingVersion v = str.skipEncapsulation();
            if (!v.Equals(Ice.Util.Encoding_1_0))
            {
                s.Write("\nencoding = ");
                s.Write(Ice.Util.encodingVersionToString(v));
            }
        }
        catch (System.IO.IOException)
        {
            Debug.Assert(false);
        }
    }

    private static byte printHeader(System.IO.StringWriter s, Ice.InputStream str)
    {
        try
        {
            str.readByte(); // Don't bother printing the magic number
            str.readByte();
            str.readByte();
            str.readByte();

            /* byte pMajor = */
            str.readByte();
            /* byte pMinor = */
            str.readByte();
            // s.Write("\nprotocol version = " + (int)pMajor + "." + (int)pMinor);

            /* byte eMajor = */
            str.readByte();
            /* byte eMinor = */
            str.readByte();
            // s.Write("\nencoding version = " + (int)eMajor + "." + (int)eMinor);

            byte type = str.readByte();
            s.Write("\nmessage type = " + (int)type + " (" + getMessageTypeAsString(type) + ')');

            byte compress = str.readByte();
            s.Write("\ncompression status = " + (int)compress + ' ');
            switch (compress)
            {
                case 0:
                {
                    s.Write("(not compressed; do not compress response, if any)");
                    break;
                }

                case 1:
                {
                    s.Write("(not compressed; compress response, if any)");
                    break;
                }

                case 2:
                {
                    s.Write("(compressed; compress response, if any)");
                    break;
                }

                default:
                {
                    s.Write("(unknown)");
                    break;
                }
            }

            int size = str.readInt();
            s.Write("\nmessage size = " + size);
            return type;
        }
        catch (System.IO.IOException)
        {
            Debug.Assert(false);
            return 0;
        }
    }

    private static byte printMessage(System.IO.StringWriter s, Ice.InputStream str, ConnectionI connection)
    {
        byte type = printHeader(s, str);

        switch (type)
        {
            case Protocol.closeConnectionMsg:
            case Protocol.validateConnectionMsg:
            {
                // We're done.
                break;
            }

            case Protocol.requestMsg:
            {
                printRequest(s, str);
                break;
            }

            case Protocol.requestBatchMsg:
            {
                printBatchRequest(s, str);
                break;
            }

            case Protocol.replyMsg:
            {
                printReply(s, str);
                break;
            }

            default:
            {
                s.Write("(unknown)");
                break;
            }
        }

        if (connection is not null)
        {
            s.Write("\ntransport = " + connection.type() + "\n");
            string connectionId = connection.endpoint().connectionId();
            if (connectionId.Length > 0)
            {
                s.Write("connection ID = " + connectionId + "\n");
            }
            s.Write(connection.ToString());
        }
        else
        {
            s.Write("\ncollocated = true");
        }

        return type;
    }

    private static string getMessageTypeAsString(byte type)
    {
        return type switch
        {
            Protocol.requestMsg => "request",
            Protocol.requestBatchMsg => "batch request",
            Protocol.replyMsg => "reply",
            Protocol.closeConnectionMsg => "close connection",
            Protocol.validateConnectionMsg => "validate connection",
            _ => "unknown",
        };
    }

    private static readonly object _globalMutex = new object();
}

File: java/src/com.zeroc.ice/src/main/java/com/zeroc/Ice/TraceUtil.java
// Copyright (c) ZeroC, Inc.

package com.zeroc.Ice;

import java.io.IOException;
import java.io.StringWriter;
import java.io.Writer;
import java.util.HashSet;
import java.util.Set;

final class TraceUtil {
    public static void traceSend(
            OutputStream str,
            Instance instance,
            ConnectionI connection,
            Logger logger,
            TraceLevels tl) {
        if (tl.protocol >= 1) {
            int p = str.pos();
            var is = new InputStream(instance, str.getEncoding(), str.getBuffer(), false);
            is.pos(0);

            StringWriter s = new StringWriter();
            byte type = printMessage(s, is, connection);

            logger.trace(
                tl.protocolCat, "sending " + getMessageTypeAsString(type) + " " + s.toString());

            str.pos(p);
        }
    }

    public static void traceRecv(
            InputStream str, ConnectionI connection, Logger logger, TraceLevels tl) {
        if (tl.protocol >= 1) {
            int p = str.pos();
            str.pos(0);

            StringWriter s = new StringWriter();
            byte type = printMessage(s, str, connection);

            logger.trace(
                tl.protocolCat,
                "received " + getMessageTypeAsString(type) + " " + s.toString());

            str.pos(p);
        }
    }

    public static void trace(
            String heading,
            InputStream str,
            ConnectionI connection,
            Logger logger,
            TraceLevels tl) {
        if (tl.protocol >= 1) {
            int p = str.pos();
            str.pos(0);

            StringWriter s = new StringWriter();
            s.write(heading);
            printMessage(s, str, connection);

            logger.trace(tl.protocolCat, s.toString());
            str.pos(p);
        }
    }

    private static final Set<String> slicingIds = new HashSet<>();

    public static synchronized void traceSlicing(
            String kind, String typeId, String slicingCat, Logger logger) {
        if (slicingIds.add(typeId)) {
            StringWriter s = new StringWriter();
            s.write("unknown " + kind + " type `" + typeId + "'");
            logger.trace(slicingCat, s.toString());
        }
    }

    private static void printIdentityFacetOperation(Writer out, InputStream stream) {
        try {
            ToStringMode toStringMode = stream.instance().toStringMode();

            Identity identity = Identity.ice_read(stream);
            out.write("\nidentity = " + Util.identityToString(identity, toStringMode));

            String[] facet = stream.readStringSeq();
            out.write("\nfacet = ");
            if (facet.length > 0) {
                out.write(StringUtil.escapeString(facet[0], "", toStringMode));
            }

            String operation = stream.readString();
            out.write("\noperation = " + operation);
        } catch (IOException ex) {
            assert false;
        }
    }

    private static void printRequest(StringWriter s, InputStream str) {
        int requestId = str.readInt();
        s.write("\nrequest id = " + requestId);
        if (requestId == 0) {
            s.write(" (oneway)");
        }

        printRequestHeader(s, str);
    }

    private static void printBatchRequest(StringWriter s, InputStream str) {
        int batchRequestNum = str.readInt();
        s.write("\nnumber of requests = " + batchRequestNum);

        for (int i = 0; i < batchRequestNum; i++) {
            s.write("\nrequest #" + i + ':');
            printRequestHeader(s, str);
        }
    }

    private static void printReply(StringWriter s, InputStream str) {
        int requestId = str.readInt();
        s.write("\nrequest id = " + requestId);

        s.write("\nreply status = ");
        // convert the signed byte into a positive int
        int replyStatusInt = str.readByte() & 0xFF;
        var replyStatus = ReplyStatus.valueOf(replyStatusInt);
        if (replyStatus != null) {
            s.write(replyStatus.toString());

            switch (replyStatus) {
                case Ok:
                case UserException:
                    EncodingVersion v = str.skipEncapsulation();
                    if (!v.equals(Util.Encoding_1_0)) {
                        s.write("\nencoding = ");
                        s.write(Util.encodingVersionToString(v));
                    }
                    break;

                case ObjectNotExist:
                case FacetNotExist:
                case OperationNotExist:
                    printIdentityFacetOperation(s, str);
                    break;

                default:
                    s.write("\nmessage = " + str.readString());
                    break;
            }
        } else {
            s.write(Integer.toString(replyStatusInt));
            // Same as default case above:
            s.write("\nmessage = " + str.readString());
        }
    }

    private static void printRequestHeader(Writer out, InputStream stream) {
        printIdentityFacetOperation(out, stream);

        try {
            byte mode = stream.readByte();
            out.write("\nmode = " + (int) mode + ' ');
            switch (OperationMode.values()[mode]) {
                case Normal -> out.write("(normal)");
                case Nonmutating -> out.write("(nonmutating)");
                case Idempotent -> out.write("(idempotent)");
                default -> out.write("(unknown)");
            }

            int sz = stream.readSize();
            out.write("\ncontext = ");
            while (sz-- > 0) {
                String key = stream.readString();
                String value = stream.readString();
                out.write(key + '/' + value);
                if (sz > 0) {
                    out.write(", ");
                }
            }

            EncodingVersion v = stream.skipEncapsulation();
            if (!v.equals(Util.Encoding_1_0)) {
                out.write("\nencoding = ");
                out.write(Util.encodingVersionToString(v));
            }
        } catch (IOException ex) {
            assert false;
        }
    }

    private static byte printHeader(Writer out, InputStream stream) {
        stream.readByte(); // Don't bother printing the magic number
        stream.readByte();
        stream.readByte();
        stream.readByte();

        //        byte pMajor = stream.readByte();
        //        byte pMinor = stream.readByte();
        //        out.write("\nprotocol version = " + (int)pMajor + "." + (int)pMinor);
        stream.readByte(); // major
        stream.readByte(); // minor

        //        byte eMajor = stream.readByte();
        //        byte eMinor = stream.readByte();
        //        out.write("\nencoding version = " + (int)eMajor + "." + (int)eMinor);
        stream.readByte(); // major
        stream.readByte(); // minor

        byte type = stream.readByte();

        try {
            out.write("\nmessage type = " + (int) type + " (" + getMessageTypeAsString(type) + ')');
            byte compress = stream.readByte();
            out.write("\ncompression status = " + (int) compress + ' ');
            switch (compress) {
                case (byte) 0 -> out.write("(not compressed; do not compress response, if any)");
                case (byte) 1 -> out.write("(not compressed; compress response, if any)");
                case (byte) 2 -> out.write("(compressed; compress response, if any)");
                default -> out.write("(unknown)");
            }

            int size = stream.readInt();
            out.write("\nmessage size = " + size);
            return type;
        } catch (IOException ex) {
            assert false;
            return 0;
        }
    }

    private static byte printMessage(StringWriter s, InputStream str, ConnectionI connection) {
        byte type = printHeader(s, str);

        switch (type) {
            case Protocol.closeConnectionMsg, Protocol.validateConnectionMsg -> { /* we're done */ }
            case Protocol.requestMsg -> printRequest(s, str);
            case Protocol.requestBatchMsg -> printBatchRequest(s, str);
            case Protocol.replyMsg -> printReply(s, str);
            default -> {}
        }

        if (connection != null) {
            s.write("\ntransport = " + connection.type() + "\n");
            String connectionId = connection.endpoint().connectionId();
            if (!connectionId.isEmpty()) {
                s.write("connection ID = " + connectionId + "\n");
            }
            s.write(connection.toString());
        } else {
            s.write("\ncollocated = true");
        }

        return type;
    }

    private static String getMessageTypeAsString(byte type) {
        return switch (type) {
            case Protocol.requestMsg -> "request";
            case Protocol.requestBatchMsg -> "batch request";
            case Protocol.replyMsg -> "reply";
            case Protocol.closeConnectionMsg -> "close connection";
            case Protocol.validateConnectionMsg -> "validate connection";
            default -> "unknown";
        };
    }

    private TraceUtil() {}
}
